quil-py/src/instruction/mod.rs

```diff
@@ -178,3 +204,56 @@ create_init_submodule! {
     ],
     errors: [ GateError, ParseMemoryReferenceError ],
 }
+
+/// Implements a copy method, as well as __copy__ and __deepcopy__ for use with Python's copy
+/// module, on any type that is a variant of the PyInstruction class.
+///
+/// Both `copy` and `__deepcopy__` are deep copies. They check the instruction for
+/// [`QubitPlaceholder`]s and create new placeholders in place of them in the copy.
+///
+/// The `__copy__` method [`Clone`]s the instruction. This means resolving [`QubitPlaceholder`]s
+/// in a copy created with Python's ``copy.copy`` will also resolve them in the original.
```

Comment:
Thoughts on this? I'm considering not providing `__copy__` at all because Python's `copy.copy(x)` should return a _shallow_ copy of x, but a `clone()` will almost always result in a deep copy.

Reply:
After chatting offline @BatmanAoD, we learned that `Clone` on a `pyo3` object will actually just return a reference, so it can be used to create a shallow copy. I've updated the tests to check that works as we expect.

I went ahead and removed the `copy` method as well since it was redundant.